File: src/main/resources/static/scripts/customize/controller/columnController.js
app.controller('columnController', function ($scope, $controller, columnService) {
    $controller('baseController', {$scope: $scope});

    /**
     * 保存或者更新操作
     */
    $scope.save = function () {
        let id = $scope.entity.id;
        if ($scope.entity.columnLevel != null) {
            if ($scope.entity.columnLevel > 1) {
                if ($scope.entity.higherId == null) {
                    alert("超过一级栏目，请选择父级栏目");
                    return
                }
            }
            let method;
            if (id != null) {
                // 如果ID不为空，则说明是更新信息操作，为空则是添加操作
                method = columnService.update($scope.entity);
            } else {
                method = columnService.save($scope.entity);
            }
            method.success(function (result) {
                $scope.myReload(result);
            });
        } else {
            alert("请选择栏目级别");
        }
    };

    /**
     * 删除操作
     * @param id 栏目ID
     */
    $scope.deleteColumn = function (id) {
        if (id != null) {
            if (confirm("确认删除栏目以及栏目下的所有文章?")) {
                columnService.delete(id).success(function (result) {
                    $scope.myReload(result);
                });
            }
        }
    };

    /**
     * 查询所有的栏目列表
     */
    $scope.levelNameList = ['column level 1', 'column level 2', 'column level 3'];
    $scope.listColumn = function () {
        columnService.listColumn().success(function (result) {
            if (result.code === 200) {
                $scope.columnList = result.data;
            }
        });
    };


    /**
     * 查询该级别下的所有栏目
     * @param level
     */
    $scope.subColumnList = [];
    $scope.listByLevel = function (level) {
        if (level != null && level > 1) {
            columnService.listByLevel(parseInt(level) - 1 + '').success(function (result) {
                if (result.code === 200) {
                    $scope.subColumnList = result.data;
                }
            });
        } else {
            $scope.subColumnList = [];
        }
    };

    /**
     * 根据ID查找该条栏目信息
     * @param id
     */
    $scope.findColumn = function (id) {
        columnService.find(id).success(function (result) {
            if (result.code === 200) {
                $scope.entity = result.data;
                $scope.listByLevel($scope.entity.columnLevel);
            }
        });
    };


    /**
     * 抽取共用块
     * @param result response result
     */
    $scope.myReload = function (result) {
        if (result.code === 200) {
            alert(result.message);
            $scope.listColumn();
        } else {
            alert(result.message);
        }
    }
});

File: src/main/java/per/neal/blog/controller/ColumnController.java
package per.neal.blog.controller;

import io.swagger.annotations.*;
import org.apache.shiro.authz.annotation.RequiresRoles;
import org.springframework.web.bind.annotation.*;
import per.neal.blog.constant.Constants;
import per.neal.blog.entity.TbColumn;
import per.neal.blog.entity.dto.R;
import per.neal.blog.service.ColumnService;

import javax.annotation.Resource;

/**
 * 栏目控制器
 *
 * @author neal
 */
@Api(tags = "栏目控制器")
@RestController
@RequestMapping("/column")
public class ColumnController {

    @ApiModelProperty("栏目service接口")
    @Resource
    private ColumnService columnService;


    @ApiOperation("新建一个栏目")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "成功保存"),
            @ApiResponse(code = Constants.ERR, message = "保存失败，返回异常信息")
    })
    @RequiresRoles(Constants.ADMIN)
    @PostMapping("/save")
    public R save(@ApiParam(name = "column", value = "TbColumn", required = true) @RequestBody TbColumn column) {
        int count = columnService.countByName(column.getColumnName().toLowerCase().trim());
        if (count > 0) {
            return R.err("该栏目名已存在,请重新填写!!!");
        }
        try {
            columnService.saveColumn(column);
            return R.ok();
        } catch (Exception e) {
            e.printStackTrace();
            return R.err(e.getLocalizedMessage());
        }
    }

    @ApiOperation("根据ID查找栏目")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "返回所有栏目", response = TbColumn.class)
    })
    @GetMapping("/findColumn")
    public R findColumn(long id) {
        return R.ok(columnService.findById(id));
    }

    @ApiOperation("查询所有的栏目")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "返回所有栏目列表", response = TbColumn.class)
    })
    @GetMapping("/listColumn")
    public R listColumn() {
        return R.ok(columnService.listColumn());
    }

    @ApiOperation("删除栏目")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "删除成功"),
            @ApiResponse(code = Constants.ERR, message = "删除失败")
    })
    @RequiresRoles(Constants.ADMIN)
    @DeleteMapping("/delete")
    public R delete(long id) {
        try {
            columnService.deleteById(id);
            return R.ok();
        } catch (Exception e) {
            e.printStackTrace();
            return R.err(e.getLocalizedMessage());
        }
    }

    @ApiOperation("修改栏目")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "修改成功"),
            @ApiResponse(code = Constants.ERR, message = "修改失败")
    })
    @RequiresRoles(Constants.ADMIN)
    @PutMapping("/modifyColumn")
    public R modifyColumn(@ApiParam(name = "column", value = "TbColumn", required = true) @RequestBody TbColumn column) {
        int count = columnService.countByName(column.getColumnName().trim());
        // 修改时，需要注意，修改的名字是否跟以前的名字一样，一样则跳过栏目名检查。
        String name = columnService.findById(column.getId()).getColumnName();
        if (!name.equalsIgnoreCase(column.getColumnName())) {
            if (count > 0) {
                return R.err("该栏目名已存在,请重新填写!!!");
            }
        }
        try {
            columnService.updateColumn(column);
            return R.ok();
        } catch (Exception e) {
            e.printStackTrace();
            return R.err(e.getLocalizedMessage());
        }
    }

    @ApiOperation("根据栏目级别查询栏目列表")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "栏目列表", response = TbColumn.class)
    })
    @GetMapping("/listByLevel")
    public R listByLevel(String columnLevel) {
        try {
            return R.ok(columnService.listByLevel(columnLevel));
        } catch (Exception e) {
            e.printStackTrace();
            return R.err(e.getLocalizedMessage());
        }
    }

    @ApiOperation("根据父级栏目ID查询所有子集栏目")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "栏目列表", response = TbColumn.class)
    })
    @GetMapping("/listByParent")
    public R listByParent(long parentId) {
        try {
            return R.ok(columnService.listByParentId(parentId));
        } catch (Exception e) {
            return R.err(e.getLocalizedMessage());
        }
    }

    @ApiOperation("栏目排行榜")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "所有栏目以及栏目文章数")
    })
    @GetMapping("/rankColumn")
    public R rankColumn() {
        return R.ok(columnService.columnRank());
    }

    @ApiOperation("栏目导航栏")
    @ApiResponses({
            @ApiResponse(code = Constants.OK, message = "栏目导航栏")
    })
    @GetMapping("/columnNav")
    public R columnNav(long id) {
        return R.ok(columnService.columnNav(id));
    }
}
